blog: add post for hackathon 2022 coffee machine (#568)
wazo-platform/wazo-platform.org: content/blog/2022-11-hackathon-coffee-machine.md

---
title: Hackathon 2022 - The Virtual Coffee Machine
date: 2023-01-19 12:00:00
author: Sébastien Duthil
category: Hackathon
tags: Web, Plugin, Coffee, Machine, App
slug: hackathon-2022-virtual-coffee-machine
status: published
---

Like every year, the Wazo hackathon was the occasion to try something new. There were various teams building in various directions, our team said "What if we made a virtual coffee machine? How far can we push Wazo to accomodate specific workflows?"

> Our team was composed of [Sylvain Boily](https://github.com/sboily), tinkerer of Doom and founder of Wazo, [Jean-François Gauthier](https://github.com/bogue), able to hide the most dreadful hacks behind delightful UIs and [Sébastien Duthil](https://github.com/sduthil), backend developer.

Reviewer: Do be shy with your description ;) I suggest this one for you, if anyone have better comments below :)

```suggestion
Our team was composed of [Sylvain Boily](https://github.com/sboily), tinkerer of Doom and founder of Wazo, [Jean-François Gauthier](https://github.com/bogue), able to hide the most dreadful hacks behind delightful UIs and [Sébastien Duthil](https://github.com/sduthil), Wazo's backend services puppet master.
```

But here's the best part: our team did not change a single line of code in the Wazo application to get this result. This page was built as a plugin to the Wazo app and the UI is hosted on a different server than the main Wazo app, and all features use the Wazo server API to get the audio working.

Here's how it works in a nutshell:

* the browser loads the Wazo app
* the browser fetches the list of enabled plugins from the Wazo server
* the browser fetches the manifests of each plugin, detailing what to add and where to get the additional UI and code from
* the browser fetches the additional UI elements and code from the server hosting the plugin
* the browser runs the plugin code and displays the plugin UI

Same thing, as a diagram:

![wazo-app-plugin.svg](../../images/blog/hackathon-2022/wazo-app-plugin.svg "Flow diagram of the previous explanation")

For the Coffee Machine we needed a few extra API in the Wazo server to add the plugin-specific logic we could not handle in the browser: music start and pause, music volume, relaying server events to a plugin-specific WebSocket. Those APIs were also added in the form of a server plugin.

> The code of the Coffee Machine plugin can be found at https://github.com/TinxHQ/hackathon-coffee. This is only an example of what Wazo can do, and keep in mind that it was written in only a few hours, so it may not be the cleanest code ever...

Reviewer: ```suggestion
The code of the Coffee Machine plugin can be found on [Github](https://github.com/TinxHQ/hackathon-coffee). This is only an example of what Wazo can do, and keep in mind that it was written in only a few hours, so it may not be the cleanest code ever...
```

> The documentation for writing plugins can be found at https://wazo-communication.github.io/euc-plugins-js-sdk/docs/web-desktop-application/

Reviewer: ```suggestion
Curious how our enterprise's SDK enable customization for our apps easily, explore our [documentation here](https://wazo-communication.github.io/euc-plugins-js-sdk/docs/web-desktop-application/)
```
